File: tests/contract/protocol/signing_test.go
package protocol_test

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"math"
	"strings"
	"testing"

	"github.com/gowebpki/jcs"
	v1 "github.com/yuanshu-ai/yuanshu/internal/protocol/v1"
)

type signingFixture struct {
	ControlDomain   string `json:"controlDomain"`
	OperationDomain string `json:"operationDomain"`
	JCSCases        []struct {
		Name          string `json:"name"`
		InputJSON     string `json:"inputJson"`
		CanonicalJSON string `json:"canonicalJson"`
	} `json:"jcsCases"`
	Control struct {
		TestOnlySeedHex           string            `json:"testOnlySeedHex"`
		PublicKeyBase64URL        string            `json:"publicKeyBase64Url"`
		Message                   v1.YuanshuMessage `json:"message"`
		CanonicalWithoutSignature string            `json:"canonicalWithoutSignature"`
		SigningInputBase64URL     string            `json:"signingInputBase64Url"`
		SignatureBase64URL        string            `json:"signatureBase64Url"`
	} `json:"control"`
	Approval struct {
		Message          v1.YuanshuMessage `json:"message"`
		CanonicalBinding string            `json:"canonicalBinding"`
		OperationDigest  string            `json:"operationDigest"`
	} `json:"approval"`
}

func TestJCSSharedVectors(t *testing.T) {
	fixture := loadSigningFixture(t)
	for _, testCase := range fixture.JCSCases {
		t.Run(testCase.Name, func(t *testing.T) {
			canonical, err := jcs.Transform([]byte(testCase.InputJSON))
			if err != nil {
				t.Fatal(err)
			}
			if string(canonical) != testCase.CanonicalJSON {
				t.Fatalf("canonical JSON = %q, want %q", canonical, testCase.CanonicalJSON)
			}
		})
	}
}

func TestControlSigningVector(t *testing.T) {
	fixture := loadSigningFixture(t)
	if fixture.ControlDomain != v1.ControlSigningDomain || fixture.OperationDomain != v1.OperationDigestDomain {
		t.Fatal("fixture domains do not match protocol constants")
	}
	original := mustJSON(t, fixture.Control.Message)
	input, err := v1.ControlSigningInput(fixture.Control.Message)
	if err != nil {
		t.Fatal(err)
	}
	if got := base64.RawURLEncoding.EncodeToString(input); got != fixture.Control.SigningInputBase64URL {
		t.Fatalf("signing input = %q, want %q", got, fixture.Control.SigningInputBase64URL)
	}
	canonical := input[len(v1.ControlSigningDomain):]
	if string(canonical) != fixture.Control.CanonicalWithoutSignature {
		t.Fatalf("canonical control = %q, want %q", canonical, fixture.Control.CanonicalWithoutSignature)
	}
	if after := mustJSON(t, fixture.Control.Message); after != original {
		t.Fatal("ControlSigningInput mutated its input")
	}

	seed, err := hex.DecodeString(fixture.Control.TestOnlySeedHex)
	if err != nil {
		t.Fatal(err)
	}
	privateKey := ed25519.NewKeyFromSeed(seed)
	publicKey := privateKey.Public().(ed25519.PublicKey)
	if got := base64.RawURLEncoding.EncodeToString(publicKey); got != fixture.Control.PublicKeyBase64URL {
		t.Fatalf("public key = %q, want %q", got, fixture.Control.PublicKeyBase64URL)
	}
	signature := ed25519.Sign(privateKey, input)
	if got := base64.RawURLEncoding.EncodeToString(signature); got != fixture.Control.SignatureBase64URL {
		t.Fatalf("signature = %q, want %q", got, fixture.Control.SignatureBase64URL)
	}
	if !ed25519.Verify(publicKey, input, signature) {
		t.Fatal("fixed Ed25519 signature did not verify")
	}
	if fixture.Control.Message.Signature == nil || *fixture.Control.Message.Signature != fixture.Control.SignatureBase64URL {
		t.Fatal("wire message does not contain the fixed signature")
	}
}

func TestControlSigningInputBindsEnvelopeAndPayload(t *testing.T) {
	fixture := loadSigningFixture(t)
	baseline, err := v1.ControlSigningInput(fixture.Control.Message)
	if err != nil {
		t.Fatal(err)
	}
	publicKeyBytes, err := base64.RawURLEncoding.DecodeString(fixture.Control.PublicKeyBase64URL)
	if err != nil {
		t.Fatal(err)
	}
	signatureBytes, err := base64.RawURLEncoding.DecodeString(fixture.Control.SignatureBase64URL)
	if err != nil {
		t.Fatal(err)
	}
	publicKey := ed25519.PublicKey(publicKeyBytes)

	mutations := map[string]func(*v1.YuanshuMessage){
		"messageId":     func(message *v1.YuanshuMessage) { message.MessageID += "_changed" },
		"type":          func(message *v1.YuanshuMessage) { message.Type = "turn.steer" },
		"sentAt":        func(message *v1.YuanshuMessage) { message.SentAt = "2026-08-01T06:00:01Z" },
		"expiresAt":     func(message *v1.YuanshuMessage) { changed := "2026-08-01T06:02:00Z"; message.ExpiresAt = &changed },
		"ownerId":       func(message *v1.YuanshuMessage) { message.OwnerID += "_changed" },
		"nodeId":        func(message *v1.YuanshuMessage) { message.NodeID += "_changed" },
		"workspaceId":   func(message *v1.YuanshuMessage) { changed := "workspace_2"; message.WorkspaceID = &changed },
		"threadId":      func(message *v1.YuanshuMessage) { changed := "thread_2"; message.ThreadID = &changed },
		"turnId":        func(message *v1.YuanshuMessage) { changed := "turn_2"; message.TurnID = &changed },
		"itemId":        func(message *v1.YuanshuMessage) { changed := "item_2"; message.ItemID = &changed },
		"streamId":      func(message *v1.YuanshuMessage) { message.StreamID += "_changed" },
		"sequence":      func(message *v1.YuanshuMessage) { message.Sequence-- },
		"correlationId": func(message *v1.YuanshuMessage) { message.CorrelationID += "_changed" },
		"nonce":         func(message *v1.YuanshuMessage) { changed := "different_nonce_1234"; message.Nonce = &changed },
		"signer":        func(message *v1.YuanshuMessage) { message.Signer.KeyID += "_changed" },
		"payload":       func(message *v1.YuanshuMessage) { message.Payload["input"] = "changed" },
	}
	for name, mutate := range mutations {
		t.Run(name, func(t *testing.T) {
			message := cloneMessage(t, fixture.Control.Message)
			mutate(&message)
			changed, err := v1.ControlSigningInput(message)
			if err != nil {
				t.Fatal(err)
			}
			if string(changed) == string(baseline) {
				t.Fatal("signed field mutation did not change signing input")
			}
			if ed25519.Verify(publicKey, changed, signatureBytes) {
				t.Fatal("fixed signature verified after a signed field mutation")
			}
		})
	}

	message := cloneMessage(t, fixture.Control.Message)
	changedSignature := strings.Repeat("B", 86)
	message.Signature = &changedSignature
	unchanged, err := v1.ControlSigningInput(message)
	if err != nil {
		t.Fatal(err)
	}
	if string(unchanged) != string(baseline) {
		t.Fatal("signature member unexpectedly changed signing input")
	}
}

func TestApprovalOperationDigestVectorAndBinding(t *testing.T) {
	fixture := loadSigningFixture(t)
	original := mustJSON(t, fixture.Approval.Message)
	digest, err := v1.ApprovalOperationDigest(fixture.Approval.Message)
	if err != nil {
		t.Fatal(err)
	}
	if digest != fixture.Approval.OperationDigest {
		t.Fatalf("operation digest = %q, want %q", digest, fixture.Approval.OperationDigest)
	}
	if after := mustJSON(t, fixture.Approval.Message); after != original {
		t.Fatal("ApprovalOperationDigest mutated its input")
	}

	transportMutations := []func(*v1.YuanshuMessage){
		func(message *v1.YuanshuMessage) { message.MessageID = "replayed_message" },
		func(message *v1.YuanshuMessage) { message.SentAt = "2026-08-01T07:00:00Z" },
		func(message *v1.YuanshuMessage) { message.StreamID = "replay_stream" },
		func(message *v1.YuanshuMessage) { message.Sequence = 999 },
		func(message *v1.YuanshuMessage) { message.CorrelationID = "replay_request" },
		func(message *v1.YuanshuMessage) { message.Payload["operationDigest"] = strings.Repeat("B", 43) },
	}
	for _, mutate := range transportMutations {
		message := cloneMessage(t, fixture.Approval.Message)
		mutate(&message)
		got, err := v1.ApprovalOperationDigest(message)
		if err != nil {
			t.Fatal(err)
		}
		if got != digest {
			t.Fatal("replay metadata changed operation digest")
		}
	}

	bindingMutations := []func(*v1.YuanshuMessage){
		func(message *v1.YuanshuMessage) { message.NodeID = "node_2" },
		func(message *v1.YuanshuMessage) { changed := "workspace_2"; message.WorkspaceID = &changed },
		func(message *v1.YuanshuMessage) { message.Payload["approvalId"] = "approval_2" },
		func(message *v1.YuanshuMessage) { message.Payload["summary"] = "Changed summary" },
		func(message *v1.YuanshuMessage) { message.Payload["futureField"] = true },
	}
	for _, mutate := range bindingMutations {
		message := cloneMessage(t, fixture.Approval.Message)
		mutate(&message)
		got, err := v1.ApprovalOperationDigest(message)
		if err != nil {
			t.Fatal(err)
		}
		if got == digest {
			t.Fatal("bound operation mutation did not change digest")
		}
	}
}

func TestSigningEncodingRejectsInvalidInput(t *testing.T) {
	fixture := loadSigningFixture(t)
	tests := map[string]func(*v1.YuanshuMessage){
		"wrong version":     func(message *v1.YuanshuMessage) { message.ProtocolVersion = "1.1" },
		"event type":        func(message *v1.YuanshuMessage) { message.Type = "turn.completed" },
		"unsafe sequence":   func(message *v1.YuanshuMessage) { message.Sequence = 9007199254740992 },
		"non-finite number": func(message *v1.YuanshuMessage) { message.Payload["input"] = math.Inf(1) },
		"invalid UTF-8":     func(message *v1.YuanshuMessage) { message.OwnerID = string([]byte{0xff}) },
		"missing signer":    func(message *v1.YuanshuMessage) { message.Signer = nil },
	}
	for name, mutate := range tests {
		t.Run(name, func(t *testing.T) {
			message := cloneMessage(t, fixture.Control.Message)
			mutate(&message)
			if _, err := v1.ControlSigningInput(message); err == nil {
				t.Fatal("expected encoding error")
			}
		})
	}

	if _, err := v1.ApprovalOperationDigest(fixture.Control.Message); err == nil {
		t.Fatal("control message was accepted as an approval event")
	}
}

func loadSigningFixture(t *testing.T) signingFixture {
	t.Helper()
	var fixture signingFixture
	readFixture(t, "fixtures/signing-vectors.json", &fixture)
	return fixture
}

func cloneMessage(t *testing.T, message v1.YuanshuMessage) v1.YuanshuMessage {
	t.Helper()
	encoded, err := json.Marshal(message)
	if err != nil {
		t.Fatal(err)
	}
	var cloned v1.YuanshuMessage
	if err := json.Unmarshal(encoded, &cloned); err != nil {
		t.Fatal(err)
	}
	return cloned
}

func mustJSON(t *testing.T, value any) string {
	t.Helper()
	encoded, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	return string(encoded)
}

File: web/src/protocol/v1/signing.test.ts
// @vitest-environment node

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import canonicalize from "canonicalize";
import { describe, expect, it } from "vitest";
import type { YuanshuMessage } from "./types.generated";
import {
  CONTROL_SIGNING_DOMAIN,
  OPERATION_DIGEST_DOMAIN,
  approvalOperationDigest,
  controlSigningInput,
} from "./signing";

interface SigningFixture {
  controlDomain: string;
  operationDomain: string;
  jcsCases: Array<{ name: string; inputJson: string; canonicalJson: string }>;
  control: {
    testOnlySeedHex: string;
    publicKeyBase64Url: string;
    message: YuanshuMessage;
    canonicalWithoutSignature: string;
    signingInputBase64Url: string;
    signatureBase64Url: string;
  };
  approval: {
    message: YuanshuMessage;
    canonicalBinding: string;
    operationDigest: string;
  };
}

const fixture = JSON.parse(
  readFileSync(resolve(process.cwd(), "..", "schemas", "yuanshu", "v1", "fixtures", "signing-vectors.json"), "utf8"),
) as SigningFixture;

describe("Protocol v1 signing encoding", () => {
  it("matches the shared RFC 8785 canonicalization vectors", () => {
    for (const testCase of fixture.jcsCases) {
      expect(canonicalize(JSON.parse(testCase.inputJson)), testCase.name).toBe(testCase.canonicalJson);
    }
  });

  it("produces the fixed control signing input and Ed25519 signature", async () => {
    expect(CONTROL_SIGNING_DOMAIN).toBe(fixture.controlDomain);
    expect(OPERATION_DIGEST_DOMAIN).toBe(fixture.operationDomain);
    const before = JSON.stringify(fixture.control.message);
    const input = controlSigningInput(fixture.control.message);
    expect(toBase64Url(input)).toBe(fixture.control.signingInputBase64Url);
    expect(new TextDecoder().decode(input.slice(new TextEncoder().encode(CONTROL_SIGNING_DOMAIN).length))).toBe(
      fixture.control.canonicalWithoutSignature,
    );
    expect(JSON.stringify(fixture.control.message)).toBe(before);

    const seed = fromHex(fixture.control.testOnlySeedHex);
    const pkcs8 = concat(fromHex("302e020100300506032b657004220420"), seed);
    const privateKey = await crypto.subtle.importKey("pkcs8", asArrayBuffer(pkcs8), { name: "Ed25519" }, false, ["sign"]);
    const signature = new Uint8Array(await crypto.subtle.sign("Ed25519", privateKey, asArrayBuffer(input)));
    expect(toBase64Url(signature)).toBe(fixture.control.signatureBase64Url);
    expect(fixture.control.message.signature).toBe(fixture.control.signatureBase64Url);

    const publicKey = await crypto.subtle.importKey(
      "raw",
      asArrayBuffer(fromBase64Url(fixture.control.publicKeyBase64Url)),
      { name: "Ed25519" },
      false,
      ["verify"],
    );
    await expect(crypto.subtle.verify("Ed25519", publicKey, asArrayBuffer(signature), asArrayBuffer(input))).resolves.toBe(true);
  });

  it("binds every control envelope group and nested payload but excludes signature", async () => {
    const baseline = controlSigningInput(fixture.control.message);
    const publicKey = await crypto.subtle.importKey(
      "raw",
      asArrayBuffer(fromBase64Url(fixture.control.publicKeyBase64Url)),
      { name: "Ed25519" },
      false,
      ["verify"],
    );
    const fixedSignature = fromBase64Url(fixture.control.signatureBase64Url);
    const mutations: Array<[string, (message: Record<string, any>) => void]> = [
      ["messageId", (message) => { message.messageId += "_changed"; }],
      ["type", (message) => { message.type = "turn.steer"; }],
      ["sentAt", (message) => { message.sentAt = "2026-08-01T06:00:01Z"; }],
      ["expiresAt", (message) => { message.expiresAt = "2026-08-01T06:02:00Z"; }],
      ["ownerId", (message) => { message.ownerId += "_changed"; }],
      ["nodeId", (message) => { message.nodeId += "_changed"; }],
      ["workspaceId", (message) => { message.workspaceId = "workspace_2"; }],
      ["threadId", (message) => { message.threadId = "thread_2"; }],
      ["turnId", (message) => { message.turnId = "turn_2"; }],
      ["itemId", (message) => { message.itemId = "item_2"; }],
      ["streamId", (message) => { message.streamId += "_changed"; }],
      ["sequence", (message) => { message.sequence -= 1; }],
      ["correlationId", (message) => { message.correlationId += "_changed"; }],
      ["nonce", (message) => { message.nonce = "different_nonce_1234"; }],
      ["signer", (message) => { message.signer.keyId += "_changed"; }],
      ["payload", (message) => { message.payload.input = "changed"; }],
    ];
    for (const [name, mutate] of mutations) {
      const message = structuredClone(fixture.control.message) as unknown as Record<string, any>;
      mutate(message);
      const changed = controlSigningInput(message as unknown as YuanshuMessage);
      expect(changed, name).not.toEqual(baseline);
      await expect(
        crypto.subtle.verify("Ed25519", publicKey, asArrayBuffer(fixedSignature), asArrayBuffer(changed)),
        name,
      ).resolves.toBe(false);
    }

    const signatureOnly = structuredClone(fixture.control.message) as unknown as Record<string, any>;
    signatureOnly.signature = "B".repeat(86);
    expect(controlSigningInput(signatureOnly as unknown as YuanshuMessage)).toEqual(baseline);
  });

  it("produces a replay-stable operation digest bound to targets and the full payload", async () => {
    const before = JSON.stringify(fixture.approval.message);
    const digest = await approvalOperationDigest(fixture.approval.message);
    expect(digest).toBe(fixture.approval.operationDigest);
    expect(JSON.stringify(fixture.approval.message)).toBe(before);

    const payload = { ...fixture.approval.message.payload };
    delete payload.operationDigest;
    const binding: Record<string, unknown> = {
      protocolVersion: fixture.approval.message.protocolVersion,
      type: fixture.approval.message.type,
      ownerId: fixture.approval.message.ownerId,
      nodeId: fixture.approval.message.nodeId,
      workspaceId: fixture.approval.message.workspaceId,
      threadId: fixture.approval.message.threadId,
      turnId: fixture.approval.message.turnId,
      itemId: fixture.approval.message.itemId,
      payload,
    };
    expect(canonicalize(binding)).toBe(fixture.approval.canonicalBinding);

    const replayMutations: Array<(message: Record<string, any>) => void> = [
      (message) => { message.messageId = "replayed_message"; },
      (message) => { message.sentAt = "2026-08-01T07:00:00Z"; },
      (message) => { message.streamId = "replay_stream"; },
      (message) => { message.sequence = 999; },
      (message) => { message.correlationId = "replay_request"; },
      (message) => { message.payload.operationDigest = "B".repeat(43); },
    ];
    for (const mutate of replayMutations) {
      const message = structuredClone(fixture.approval.message) as unknown as Record<string, any>;
      mutate(message);
      await expect(approvalOperationDigest(message as unknown as YuanshuMessage)).resolves.toBe(digest);
    }

    const bindingMutations: Array<(message: Record<string, any>) => void> = [
      (message) => { message.nodeId = "node_2"; },
      (message) => { message.workspaceId = "workspace_2"; },
      (message) => { message.payload.approvalId = "approval_2"; },
      (message) => { message.payload.summary = "Changed summary"; },
      (message) => { message.payload.futureField = true; },
    ];
    for (const mutate of bindingMutations) {
      const message = structuredClone(fixture.approval.message) as unknown as Record<string, any>;
      mutate(message);
      await expect(approvalOperationDigest(message as unknown as YuanshuMessage)).resolves.not.toBe(digest);
    }
  });

  it("rejects unsupported and non-I-JSON inputs", async () => {
    const wrongVersion = structuredClone(fixture.control.message) as unknown as Record<string, any>;
    wrongVersion.protocolVersion = "1.1";
    expect(() => controlSigningInput(wrongVersion as unknown as YuanshuMessage)).toThrow(/unsupported/);

    const nonFinite = structuredClone(fixture.control.message) as unknown as Record<string, any>;
    nonFinite.payload.input = Number.POSITIVE_INFINITY;
    expect(() => controlSigningInput(nonFinite as unknown as YuanshuMessage)).toThrow(/finite/);

    const badUnicode = structuredClone(fixture.control.message) as unknown as Record<string, any>;
    badUnicode.ownerId = "\ud800";
    expect(() => controlSigningInput(badUnicode as unknown as YuanshuMessage)).toThrow(/surrogate/);

    await expect(approvalOperationDigest(fixture.control.message)).rejects.toThrow(/approval.requested/);
  });
});

function fromHex(value: string): Uint8Array {
  return Uint8Array.from(value.match(/../g) ?? [], (byte) => Number.parseInt(byte, 16));
}

function fromBase64Url(value: string): Uint8Array {
  return new Uint8Array(Buffer.from(value, "base64url"));
}

function toBase64Url(value: Uint8Array): string {
  return Buffer.from(value).toString("base64url");
}

function concat(first: Uint8Array, second: Uint8Array): Uint8Array {
  const result = new Uint8Array(first.length + second.length);
  result.set(first);
  result.set(second, first.length);
  return result;
}

function asArrayBuffer(value: Uint8Array): ArrayBuffer {
  return value.slice().buffer as ArrayBuffer;
}
